[exporterhelper]: Add RequestMiddleware extension interface

raghu999 opened the pull request:
### Description

This PR introduces the `RequestMiddleware` interface and integrates it into the `exporterhelper`. This is a generalization of the previously proposed `ConcurrencyController`, allowing for broader control over request execution.

**Changes:**

* Defines the `RequestMiddleware` interface in the `exporter/exporterhelper/xexporterhelper` package.
* Updates the `exporterhelper`'s sending queue to accept a list of `request_middlewares`.
* Delegates request execution logic to these middlewares, allowing extensions (such as a dynamic concurrency controller) to intercept and manage export requests.

### Link to tracking issue

Relates to #14080 (Note: This PR is a prerequisite required for fixing #14080)

### Testing

* Added unit tests for the new `RequestMiddleware` interface and its integration with the queue sender.
* Verified that existing `exporterhelper` tests pass to ensure no regression in current behavior.

### Documentation

* Added GoDoc comments for the new interface and methods.

raghu999 committed:
feat: Add ConcurrencyController interface for ARC and exporterhelper plumbing

raghu999 committed:
feat: Add ConcurrencyController interface for ARC and exporterhelper plumbing

raghu999 committed:
update go.mod

raghu999 committed:
update go.mod

raghu999 committed:
Merge branch 'controller-interface' of github.com-raghu999:raghu999/opentelemetry-collector into controller-interface

raghu999 committed:
fix: queue sender

raghu999 committed:
make ARC controller opt-in and non-invasive and keep default queue behavior unchanged when ARC is disabled

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 commented:
@axw @dmitryax @bogdandrutu gentle ping on this PR (#14318). I wanted to check if you could take a look when you have a moment.

This introduces the `ConcurrencyController` interface + minimal exporterhelper plumbing to allow dynamic concurrency control (intended for the ARC extension path), with unit tests included. The change should be non-invasive unless a controller is explicitly configured.

Would appreciate a review on the API shape/placement and the integration points. Happy to iterate quickly on any feedback or adjust/split if needed. Thanks!

axw commented on extension/extensioncapabilities/interfaces.go:
```diff
@@ -47,3 +49,29 @@ type ConfigWatcher interface {
+	// CreateConcurrencyController creates a controller for a specific exporter and signal.
+	// This allows the extension to maintain separate limits per exporter/signal if desired.
+	CreateConcurrencyController(component.ID, pipeline.Signal) (ConcurrencyController, error)
+}
```
Did you consider making the interface a bit more general, to be a sort of middleware interface for exporter requests? Something like:

```go
import (
	"go.opentelemetry.io/collector/component"
	"go.opentelemetry.io/collector/exporter/exporterhelper/internal/request"
	"go.opentelemetry.io/collector/exporter/exporterhelper/internal/sender"
	"go.opentelemetry.io/collector/pipeline"
)

type RequestMiddleware interface {
	WrapSender(set RequestMiddlewareSettings, next sender.Sender[request.Request]) (sender.Sender[request.Request], error)
}

type RequestMiddlewareSettings struct {
	Signal        pipeline.Signal
	ID               component.ID
	Telemetry component.TelemetrySettings
}
```

Then the acquire, release, and timing bits -- which are specific to ARC -- can be encapsulated.

axw reviewed (commented):
Thanks @raghu999!

My main feedback so far is:
 - The interface is very narrow, and seems a bit too coupled to ARC. I'd like to see if we can change that into a more general exporter request/sender middleware interface without compromising ARC.
 - The minConsumersWithController bit feels off. I'm not convinced we should change the default due to some other setting - seems like that would be surprising. A couple of thoughts:
     - Can we for now just have the extension log a warning if the default is set low?
     - Would it make sense for the controller to be able to add consumers? Maybe as a separate extension point in asyncQueue, but referencing the same extension?

raghu999 committed:
Incorporate review comments add the middleware interface

raghu999 commented on exporter/exporterhelper/README.md:
The adaptive concurrency extension implementation is already ready on my side, which is why I went ahead and documented the new sending_queue.concurrency_controller knob and interface now.
If you’d prefer, I can drop the README changes from this PR and send them along with the follow-up/stacked PR that adds the adaptive concurrency extension implementation.

raghu999 committed:
Removed if qs.mw != nil checks in Start and Shutdown and Changed assert.Equal to assert.GreaterOrEqual

axw commented on exporter/exporterhelper/internal/queuebatch/config.go:
```diff
@@ -37,6 +37,11 @@ type Config struct {
+	// RequestMiddlewareID if not nil, enables a request middleware extension (e.g. adaptive concurrency)
+	// that intercepts export requests after they are pulled from the queue.
+	// Note: Config key remains "concurrency_controller" for consistency with initial design/docs.
+	RequestMiddlewareID *component.ID `mapstructure:"concurrency_controller"`
```
```suggestion
	RequestMiddlewares *component.ID `mapstructure:"request_middlewares"`
```

The YAML tag is wrong. Also, this should probably be a list, like confighttp's "middlewares". I don't love this name I'm suggesting, but it's at least consistent.

axw reviewed (commented):
> I refactored the ARC-coupled interface into a generic RequestMiddleware / RequestMiddlewareFactory. I explored the WrapSender(... internal/request,sender ...) style, but extensioncapabilities can’t depend on exporterhelper/internal/... types due to Go internal visibility rules and it would also introduce an import cycle. Using Handle(ctx, next func(ctx) error) keeps the interface general and decoupled while letting extensions encapsulate timing/permits/error logic.

The interface doesn't have to live in `extensioncapabilities`. For example, there's https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/extensionmiddleware for HTTP and gRPC middleware. I wouldn't recommend adding it in there, just using it as an example. Perhaps we could introduce a new package under https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/xextension?

raghu999 committed:
Incorporate review suggestions

raghu999 commented:
> > I refactored the ARC-coupled interface into a generic RequestMiddleware / RequestMiddlewareFactory. I explored the WrapSender(... internal/request,sender ...) style, but extensioncapabilities can’t depend on exporterhelper/internal/... types due to Go internal visibility rules and it would also introduce an import cycle. Using Handle(ctx, next func(ctx) error) keeps the interface general and decoupled while letting extensions encapsulate timing/permits/error logic.
> 
> The interface doesn't have to live in `extensioncapabilities`. For example, there's https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/extensionmiddleware for HTTP and gRPC middleware. I wouldn't recommend adding it in there, just using it as an example. Perhaps we could introduce a new package under https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/xextension?

Thanks for the review @axw! I've updated the PR to incorporate all suggested changes:

Configuration (config.go):

1. Renamed Field: Changed RequestMiddlewareID to RequestMiddlewares.
2. Updated Type: Changed the type to a list ([]component.ID) to be consistent with confighttp and allow multiple middlewares.
3. Updated Tag: Switched the YAML tag to mapstructure:"request_middlewares".
4. Documentation: Updated the code comments to reflect the list type and removed the concurrency controller documentation 

Interface Location:

- Refactoring: As suggested, I removed the RequestMiddleware and RequestMiddlewareFactory interfaces from extensioncapabilities.

- New Location: I've moved them to a new package go.opentelemetry.io/collector/extension/xextension/extensionmiddleware. This keeps the experimental middleware capabilities separate from the stable core extension interfaces.

Ready for a re-review!

raghu999 committed:
Merge branch 'main' into controller-interface

axw commented on extension/xextension/extensionmiddleware/interfaces.go:
```diff
@@ -0,0 +1,32 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package extensionmiddleware // import "go.opentelemetry.io/collector/extension/xextension/extensionmiddleware"
```
I'd still like to see the interface I described in https://github.com/open-telemetry/opentelemetry-collector/pull/14318#discussion_r2663052408

In hindsight, suggesting xextension was a mistake, since it wouldn't then be possible to implement that interface as the request/sender types are under `exporterexporterhelper/internal`. I played with this locally, and I think this should work:

 - Define the interface type in `exporter/exporterhelper/internal`
 - Define a type alias in `exporter/exporterhelper/xexporterhelper`

axw commented on exporter/exporterhelper/internal/queue_sender.go:
```diff
@@ -36,23 +43,159 @@ func NewDefaultQueueConfig() queuebatch.Config {
 	exportFailureMessage string,
 	next sender.Sender[request.Request],
 ) (sender.Sender[request.Request], error) {
+	if len(qCfg.RequestMiddlewares) == 0 {
```
I think all the code can be simplified if:
 - NewQueueSender just stores "next" in a new field, and references that field rather than the parameter in exportFunc
 - queueSender.Start overrides the next field with the wrapped sender

raghu999 committed:
Move controller interface to exporter helper internal

axw commented on exporter/exporterhelper/internal/queue_sender.go:
```diff
@@ -36,23 +35,162 @@ func NewDefaultQueueConfig() queuebatch.Config {
+
+// Start overrides the default Start to resolve the extensions and wrap the sender.
+func (qs *queueSender) Start(ctx context.Context, host component.Host) error {
+	mws := make([]requestmiddleware.RequestMiddleware, 0, len(qs.mwIDs))
```
I don't see the requestmiddleware code - seems you missed adding it to your commits?

raghu999 replied:
@axw  Apologies, yes, looks like I missed adding the requestmiddleware code. Just added, please check. I also incorporated the other review comments (removed the warning logic and the superfluous tests).

axw commented on exporter/exporterhelper/internal/queue_sender_test.go:
```diff
@@ -59,3 +61,338 @@ func TestQueueConfig_Validate(t *testing.T) {
+	}
+}
+
+func TestQueueSender_RequestMiddleware_Disabled_NoInteraction(t *testing.T) {
```
This test seems superfluous. It's not just "disabled", it's not referenced at all. I think we can drop this.

axw commented on exporter/exporterhelper/internal/queue_sender.go:
```diff
@@ -36,23 +35,162 @@ func NewDefaultQueueConfig() queuebatch.Config {
+		)
+	}
+}
+
```
The warning only makes sense if the middleware includes a concurrency controller. Can you please remove this, and we can see if there's a good way to reintroduce it when adding the ARC extension? e.g. maybe we can provide the NumConsumers value to the extension when starting it.

raghu999 committed:
fix: updated request.go

raghu999 committed:
feat: added request middleware

raghu999 committed:
Incorporate review comments

axw reviewed (commented):
Thanks @raghu999, it looks pretty clean to me now. I'm not sure what the Sizer change is about - that seems off.

I think it's ready for a second opinion now. @dmitryax do you have opinions on this?

raghu999 committed:
Update exporter/exporterhelper/internal/queue_sender.go

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

raghu999 committed:
Merge branch 'main' into controller-interface

axw commented:
@raghu999 I see, doesn't seem to be implement that method outside of exporterhelper at the moment.

But... why do you need to do that in this PR? All you should need to do is implement the Sender, and pass through the Request?

raghu999 committed:
Merge branch 'main' into controller-interface

axw commented:
> The reason I need the alias in this PR is for unit testing the middleware. To write a test like TestWrapSender, I need to create a mock Request implementation to pass into Send(). Because Request is an interface, my mock struct must implement all its methods, including MergeSplit.

Something like this should be good enough I think?

```go
type mockRequest struct {
    xexporterhelper.Request
}
```

If you were to invoke any methods on the interface it would panic, but since you don't need to that's OK.

raghu999 committed:
fix: updated request.go

raghu999 commented:
> > The reason I need the alias in this PR is for unit testing the middleware. To write a test like TestWrapSender, I need to create a mock Request implementation to pass into Send(). Because Request is an interface, my mock struct must implement all its methods, including MergeSplit.
> 
> Something like this should be good enough I think?
> 
> ```go
> type mockRequest struct {
>     xexporterhelper.Request
> }
> ```
> 
> If you were to invoke any methods on the interface it would panic, but since you don't need to that's OK.

@axw Great suggestion! I missed that embedding the interface would satisfy the compiler requirements for the unused methods.

That solves the issue cleanly without needing to export SizerType. I have updated the test to use the embedded interface pattern and removed the SizerType alias from this PR. Thanks!

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
chore: update changelog and fix lint

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

bogdandrutu reviewed (changes requested):
I am failing to understand this (sorry not as smart). But why not having the server push back with a "retry-after"? Based on my limited knowledge is almost impossible for you to solve this at the client side, you need the server to push back on the requests and client respect that independent of how many consumers you have (e.g. consumers can be increased also by having more sources of data and you cannot control across sources).

Also, we need to understand and separ the consumption from the queue and the sending part (if needed). I would like to better understand why retry-after mechanism does not work since that is the recommended HTTP(and gRPC) way of dealing with this.

raghu999 committed:
fix: update exporterhelper

raghu999 commented:
@bogdandrutu, thank you for the feedback. You are correct that retry-after is the standard for reactive backpressure; however, this PR introduces the RequestMiddleware interface to support proactive client-side management like Adaptive Concurrency Control (ARC).

While retry-after triggers once a server is already saturated, ARC monitors latency trends to adjust concurrency before the server drops requests or experiences resource exhaustion. By utilizing the WrapSender pattern, we decouple this logic from the core exporterhelper, keeping it lean while allowing advanced extensions to hook into the request lifecycle to maintain optimal throughput

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 commented:
@bogdandrutu @dmitryax @axw  we'd love to get your eyes on the ARC implementation strategies we've proposed here. We want to ensure this aligns perfectly with the Collector’s long-term roadmap.

We are happy to pivot based on your feedback whether that’s refining the current PR or moving this to a dedicated contrib component. My Company is committed to maintaining the ARC extension or ARC exporter helper as a core part of our observability stack. Please let us know how you'd like us to proceed!

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 commented:
@bogdandrutu Any feedback on this?

raghu999 committed:
Merge branch 'main' into controller-interface

meridianmindx commented:
This is an interesting architectural pattern — introducing a `RequestMiddleware` interface to generalize control over request execution in the exporter helper. The approach cleanly separates cross-cutting concerns like dynamic concurrency control.

A few questions for consideration:
- Are there plans for built-in middleware implementations (e.g., circuit breaker, adaptive batch size) that would ship with the collector?
- How does middleware ordering work? Is there a guarantee about execution order when multiple middlewares are registered?
- Does the interface allow mutating both the request and the context, or just wrapping execution?

This pattern could also be useful for other components beyond exporters (receivers, processors). Thanks for the clean abstraction!

raghu999 commented:
Thanks everyone for the continued feedback, reviews, and patience! I want to provide a quick update to address recent questions and clarify the architectural direction of the extension.

**@axw** Apologies for the delay in following up! Thanks again for the review and approval. While we coordinate the best time to sync up at an upcoming SIG meeting, I wanted to lay out the technical details here so we can get a head start asynchronously.

**@bogdandrutu** Gentle ping to keep this on your radar! To address the lingering architectural concerns regarding the algorithm, I want to provide some context on the design.

This approach is heavily influenced by the Adaptive Request Concurrency (ARC) mechanisms used by Vector, as well as Netflix's "Performance Under Load" architecture. The primary goal is to completely eliminate the need for static rate limits that require constant manual tuning, and instead automatically find the optimal maximum throughput. 

**How the internal mechanisms achieve this:**
* **AIMD & EWMA Control Law:** The `Controller` relies on an Additive Increase / Multiplicative Decrease (AIMD) algorithm. To evaluate downstream health, it tracks the Round Trip Time (RTT) of requests and calculates a healthy latency baseline using an Exponentially Weighted Moving Average (EWMA).
* **Dynamic Backpressure:** The extension automatically throttles the concurrency limit when it detects explicit backpressure (e.g., retryable errors) or when the recent RTT exceeds the calculated threshold, indicating a latency spike.
* **Optimized Permit Gating:** To ensure concurrency management doesn't introduce unwanted overhead, active parallel requests are gated by a custom `TokenPool`. We prioritized lean, optimized code for high-throughput performance by implementing a fast path in the pool's `Acquire` method to completely bypass slow-path allocations.

---

###  RequestMiddleware Abstraction

**@meridianmindx** Thanks for the thoughtful review and excellent questions regarding the abstraction!

Here is a breakdown of how the middleware is designed to function:
1.  **Built-in implementations:** While ARC is the immediate driver for this interface, standardizing circuit breaking, dynamic batch sizing, and advanced rate limiting are absolutely the logical next steps once this foundation is merged.
2.  **Ordering:** Middlewares will be executed sequentially in the exact order they are configured and registered in the slice, forming a standard chain.
3.  **Mutation:** The interface is designed to wrap request execution. This allows you to pass down a modified `context.Context` (useful for timeouts or tracing) and directly control the flow of the request.
4.  **Beyond exporters:** Spot on! The adaptive concurrency mechanism is inherently designed so it can also act as an HTTP/gRPC server-side interceptor to protect the Collector's ingress (receivers), not just the exporters.

---

Let me know if this helps clarify the design approach for the interface and algorithm. I'm happy to iterate further right here on the PR if there are specific adjustments you'd like to see!

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 commented:
**@bogdandrutu**  It is a completely valid question and a very common architectural debate when dealing with distributed systems. 

You are absolutely right that `Retry-After` is the standard and recommended mechanism for server-side backpressure. However, relying *exclusively* on `Retry-After` has a fundamental limitation for high-throughput observability pipelines: it is purely **reactive**.

By the time a backend (like Elasticsearch or an OTLP gateway) is issuing 429s or 503s with `Retry-After` headers, it is already in a state of distress. The server is actively burning CPU cycles to accept the connection, read the headers, determine it is overwhelmed, and format a rejection. In addition, network bandwidth is wasted transmitting payloads that are immediately dropped.

Here is why a client-side Adaptive Request Concurrency (ARC) mechanism is not only possible, but critical to solving this:

**1. RTT as a Universal Shared Signal (Addressing the Cross-Client Issue)**
You correctly pointed out that one Collector instance doesn't know about the traffic generated by other instances. That is actually the exact reason ARC works so well! The extension uses Round Trip Time (RTT) as its primary health indicator. 
If 50 different Collector instances suddenly burst traffic to the same backend, the backend's queues will fill and its latency will naturally increase. All 50 independent ARC controllers will detect this RTT degradation simultaneously and independently back off *before* the server is forced to issue a 429. It leverages the exact same principles as TCP congestion control. 

**2. Proactive (ARC) vs. Reactive (Retry-After)**
`Retry-After` relies on hitting a wall. It creates a "sawtooth" pattern of traffic: burst -> overwhelm the server -> get 429s -> stop -> burst again when the timer expires (which often causes a thundering herd problem).
ARC, via its AIMD control law and EWMA latency tracking, finds the "sweet spot" of maximum throughput just below the server's breaking point and dynamically hovers there. 

**3. Separating the Queue from the Sender**
The `RequestMiddleware` abstraction actually facilitates exactly what you are asking for separating the queue consumption from the network sending. 
Currently, if we set `num_consumers: 10`, we artificially cap our throughput even if the downstream is completely idle. With ARC, we can set `num_consumers: 200` (allowing the queue to drain rapidly and utilize CPU efficiently) but let the middleware dynamically limit the *actual in-flight HTTP/gRPC requests* to what the network can currently sustain.

Think of `Retry-After` as the airbag, and ARC as the anti-lock brakes. We absolutely still want the server to send `Retry-After` when necessary (and ARC will immediately cut concurrency when it sees retryable errors!), but ARC's job is to prevent us from crashing into that wall in the first place.

Let me know if this helps clarify the philosophy behind the client-side approach!

raghu999 commented:
@bogdandrutu Gentle ping on this. We've been waiting on your feedback regarding the requested changes for the last few months. Could you please review when you have a moment so we can figure out the best path forward? Appreciate your time!

raghu999 commented:
@axw @dmitryax @dmathieu @meridianmindx 
Hey team, I’d like to get your advice on the best protocol for moving forward here. Since the PR has been blocked by the requested changes for a few months without a follow-up response, I want to make sure I'm following the right community process.

How would the maintainers like me to proceed?
- Should I continue to hold off and wait for @bogdandrutu to have the bandwidth to review the responses?

- Would it be cleaner to close this and submit a new PR to reset the review state and get fresh eyes on it?

- Is there a different architectural implementation for ARC that the community would prefer I explore instead of the RequestMiddleware approach?

I'm eager to unblock this phase of the pipeline and would really appreciate your guidance on how to navigate this so we can keep these contributions moving.

jmacd commented on exporter/exporterhelper/xexporterhelper/request_middleware.go:
```diff
@@ -0,0 +1,25 @@
+	// OnRequest is called before the request is sent.
+	// It can block to implement concurrency control or return an error to cancel the request.
+	OnRequest(ctx context.Context) (func(error), error)
+}
```
Is this used? I'm confused by two RequestMiddleware extensions, one here and one in requestmiddleware/interfaces.go

raghu999 replied:
Good catch this is confusing as written.

The reason this was introduced is to support ARC as an extension rather than baking ARC-specific logic directly into exporterhelper. In the earlier ARC PR, the feedback was that the ARC algorithm is too opinionated to live inside exporterhelper, and that exporterhelper should instead expose a generic extension point for wrapping exporter request execution.

That is what this PR is trying to provide: a generic request middleware hook that exporterhelper can use to let extensions control request execution, with ARC being the first intended use case.

However, I agree the current code makes that unclear because there are effectively two RequestMiddleware shapes:

- the internal requestmiddleware.RequestMiddleware, which is what exporterhelper actually wires into the send pipeline
- the public xexporterhelper.RequestMiddleware, which was meant to be the extension-facing API but currently has a different OnRequest(ctx) shape

I’ll clean this up so there is only one effective middleware contract. The public xexporterhelper.RequestMiddleware should re-export or alias the same middleware shape used internally by exporterhelper, rather than defining a separate OnRequest contract. That should make it clear that the public API exists so future extensions like ARC can plug into exporterhelper without adding ARC-specific logic to exporterhelper itself.

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
This commit is introducing the RequestMiddleware open interface following
the merged component-interface RFC, correcting all the review findings
raised by @jmacd and @axw, and adding the benchmark evidence table that
was promised to @dmitryax in issue #14080.

Interface changes (RFC compliance):

- Removed the duplicate RequestMiddleware interface from the internal
  requestmiddleware package which was causing confusion since two symbols
  with the same name existed side-by-side. The package now only exports
  RequestMiddlewareSettings. The queue_sender uses a local unexported
  senderWrapper interface for the type assertion and calls out explicitly
  why structural typing satisfies xexporterhelper.RequestMiddleware.

- Removed dead middleware field and its initialisation block from
  base_exporter.go. This code was never populated and was leftover from
  an earlier design iteration.

- Added missing compile-time assertion var _ RequestMiddleware =
  (*WrapSenderFunc)(nil) in xexporterhelper as required by the RFC for
  every open interface having a companion function type.

- Added featuregate as a direct dependency in go.mod.

Test helpers:

- Added xexporterhelpertest submodule with NewNop() and NewErr()
  following the extensionmiddlewaretest pattern per RFC requirement.

ARC extension:

- Corrected GetHTTPHandler and GetGRPCServerOptions signatures to
  properly implement extensionmiddleware.HTTPServer and GRPCServer.
  Previously the method bodies were correct but the signatures did not
  match the interface (missing context.Context, wrong return type for
  GetHTTPHandler).

- Updated go.mod with full local replace directives and ran go mod tidy.

- Add xexporterhelpertest to versions.yaml (was missing, breaks CI)
- Restore correct import grouping in generated_telemetrytest.go
- Rename gate to pkg.exporterhelper.RequestMiddleware per mdatagen
  convention; drop skip_strict_validation and the wrong rationale
- Fix from_version: gate is new, v0.150.0 never had it
- Trim changelog to what actually changed in this PR
- Document WrapSender lifecycle contract: returned sender must not
  cascade Start/Shutdown into next
- Move BenchmarkQueueSender_ARC to the ARC extension; trim
  queue_sender_arc_test.go back to original scope
- Guard feature gate in Start() not just Validate()
- Revert unrelated README capitalisation change

raghu999 committed:
This commit is introducing the RequestMiddleware open interface following
the merged component-interface RFC, correcting all the review findings
raised by @jmacd and @axw, and adding the benchmark evidence table that
was promised to @dmitryax in issue #14080.

Interface changes (RFC compliance):

- Removed the duplicate RequestMiddleware interface from the internal
  requestmiddleware package which was causing confusion since two symbols
  with the same name existed side-by-side. The package now only exports
  RequestMiddlewareSettings. The queue_sender uses a local unexported
  senderWrapper interface for the type assertion and calls out explicitly
  why structural typing satisfies xexporterhelper.RequestMiddleware.

- Removed dead middleware field and its initialisation block from
  base_exporter.go. This code was never populated and was leftover from
  an earlier design iteration.

- Added missing compile-time assertion var _ RequestMiddleware =
  (*WrapSenderFunc)(nil) in xexporterhelper as required by the RFC for
  every open interface having a companion function type.

- Added featuregate as a direct dependency in go.mod.

Test helpers:

- Added xexporterhelpertest submodule with NewNop() and NewErr()
  following the extensionmiddlewaretest pattern per RFC requirement.

ARC extension:

- Corrected GetHTTPHandler and GetGRPCServerOptions signatures to
  properly implement extensionmiddleware.HTTPServer and GRPCServer.
  Previously the method bodies were correct but the signatures did not
  match the interface (missing context.Context, wrong return type for
  GetHTTPHandler).

- Updated go.mod with full local replace directives and ran go mod tidy.

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 commented:
@jmacd @axw @dmathieu @dmitryax — thank you all for the detailed review feedback and for pointing out the confusion around the two `RequestMiddleware` symbols. Your observation was spot on, and I have taken care of all the points raised, including bringing the implementation in line with the recently merged component-interface RFC at https://github.com/open-telemetry/opentelemetry-collector/pull/14532.

Kindly find below a summary of all the changes done in this update.

**Removal of the duplicate interface**

The internal `requestmiddleware/interfaces.go` was having a `RequestMiddleware` interface defined alongside the public one in `xexporterhelper`. That is the exact duplication @jmacd had flagged. That internal interface has been removed entirely. The file is now only keeping `RequestMiddlewareSettings` — the shared settings struct that is required by both the internal queue sender and the public `xexporterhelper` package.

**Local unexported interface for type assertion**

In `queue_sender.go`, the type assertion is now done against a local unexported `senderWrapper` interface instead of against the old internal `RequestMiddleware`. A comment has been added on that interface explaining why Go's structural typing is guaranteeing that any extension implementing the public `xexporterhelper.RequestMiddleware` will also be satisfying this local interface automatically. This way there is only one externally visible symbol called `RequestMiddleware` in the entire codebase.

**Single canonical interface in xexporterhelper**

`xexporterhelper/request_middleware.go` is now the single source of truth for the `RequestMiddleware` interface. The compile-time assertion `var _ RequestMiddleware = (*WrapSenderFunc)(nil)` has also been added as the RFC is mandating for every open interface that has a companion function type.

**Test helpers subpackage**

A new `xexporterhelpertest` submodule has been added with `NewNop()` and `NewErr()` constructors, following exactly the same pattern as `extensionmiddlewaretest`. The RFC is requiring every public interface package to be having a companion test helper module, and the same has been done here.

**Removal of dead code from base_exporter.go**

The `middleware` field and its initialisation block have been removed from `base_exporter.go`. That code was from an earlier design iteration and was never actually being populated anywhere. It was causing unnecessary confusion.

I believe the above changes are addressing all the review concerns. Kindly have a look at the updated diff and do let me know if there is anything further requiring attention.

raghu999 commented:
@dmitryax @jmacd  I have added the benchmark findings here https://github.com/open-telemetry/opentelemetry-collector/issues/14080#issuecomment-4733598654

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Address review feedback on RequestMiddleware PR (#14318)

Fix all issues raised in review:

- Run go mod tidy in xexporterhelpertest after backoff v5 -> v6 upgrade
- Register xexporterhelpertest in versions.yaml (missing, broke CI)
- Restore correct import grouping in generated_package_test.go and
  generated_telemetrytest.go (stdlib -> third-party -> local collector)
- Rename feature gate to pkg.exporterhelper.RequestMiddleware per mdatagen
  naming convention; drop skip_strict_validation and wrong rationale
- Set from_version to v0.155.0 (gate is new, v0.150.0 was incorrect)
- Trim changelog to what this PR actually changes
- Document WrapSender lifecycle contract: returned sender must not
  cascade Start/Shutdown into next
- Move BenchmarkQueueSender_ARC to ARC extension; trim
  queue_sender_arc_test.go back to original 263 lines
- Enforce feature gate in Start() not just Validate()
- Revert unrelated README capitalisation change on line 24
- Simplify AI-sounding code comments to read naturally

raghu999 committed:
Merge branch 'main' into controller-interface

jmacd reviewed (approved):
Looks good thank you @raghu999

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface

raghu999 committed:
Merge branch 'main' into controller-interface
